src/validation/func.rs

```diff
@@ -99,397 +223,1071 @@ impl Validator {
+				return Ok(Outcome::Unreachable);
+			}
+			BrIf(depth) => {
+				Validator::validate_br_if(context, depth)?;
+
+				let target = require_target(
+					depth,
+					&context.value_stack,
+					&context.frame_stack,
+				);
+				context.sink.emit_br_nez(target);
+			}
+			BrTable(ref table, default) => {
+				Validator::validate_br_table(context, table, default)?;
+
+				let mut targets = Vec::new();
+				for depth in table.iter() {
+					let target = require_target(
+						*depth,
+						&context.value_stack,
+						&context.frame_stack,
+					);
+					targets.push(target);
+				}
+				let default_target = require_target(
+					default,
+					&context.value_stack,
+					&context.frame_stack,
+				);
+				context.sink.emit_br_table(&targets, default_target);
+
+				return Ok(Outcome::Unreachable);
+			}
+			Return => {
+				if let BlockType::Value(value_type) = context.return_type()? {
+					tee_value(&mut context.value_stack, &context.frame_stack, value_type.into())?;
+				}
+
+				let drop_keep = drop_keep_return(
+					&context.locals,
+					&context.value_stack,
+					&context.frame_stack
+				);
+				context.sink.emit(isa::Instruction::Return(drop_keep));
+
+				return Ok(Outcome::Unreachable);
+			}
+
+			Call(index) => {
+				Validator::validate_call(context, index)?;
+				context.sink.emit(isa::Instruction::Call(index));
+			}
+			CallIndirect(index, _reserved) => {
+				Validator::validate_call_indirect(context, index)?;
+				context.sink.emit(isa::Instruction::CallIndirect(index));
+			}
+
+			Drop => {
+				Validator::validate_drop(context)?;
+				context.sink.emit(isa::Instruction::Drop);
+			}
+			Select => {
+				Validator::validate_select(context)?;
+				context.sink.emit(isa::Instruction::Select);
+			}
+
+			GetLocal(index) => {
+				// We need to calculate relative depth before validation since
+				// it will change the value stack size.
+				let depth = relative_local_depth(
+					index,
+					&context.locals,
+					&context.value_stack,
+				)?;
+				Validator::validate_get_local(context, index)?;
+				context.sink.emit(
+					isa::Instruction::GetLocal(depth),
+				);
+			}
+			SetLocal(index) => {
+				Validator::validate_set_local(context, index)?;
+				let depth = relative_local_depth(
+					index,
+					&context.locals,
+					&context.value_stack,
+				)?;
+				context.sink.emit(
+					isa::Instruction::SetLocal(depth),
+				);
+			}
+			TeeLocal(index) => {
+				Validator::validate_tee_local(context, index)?;
+				let depth = relative_local_depth(
+					index,
+					&context.locals,
+					&context.value_stack,
+				)?;
+				context.sink.emit(
+					isa::Instruction::TeeLocal(depth),
+				);
+			}
+			GetGlobal(index) => {
+				Validator::validate_get_global(context, index)?;
+				context.sink.emit(isa::Instruction::GetGlobal(index));
+			}
+			SetGlobal(index) => {
+				Validator::validate_set_global(context, index)?;
+				context.sink.emit(isa::Instruction::SetGlobal(index));
+			}
+
+			I32Load(align, offset) => {
```

Comment:
Could you generate this with a macro? It seems error-prone.

Reply:
Yeah, you're right but I would rather not touch this code! All instructions, which happened to be placed after `TeeLocal`, were mechanically and carefully edited with the help of multiline cursor feature :D

Reply:
If this code ever needs to be changed we have to convert it to be a macro. Will tests fail if we mess it up?

Reply:
> If this code ever needs to be changed we have to convert it to be a macro.

Agree. 

> Will tests fail if we mess it up?

I guess so